text_generation/causal_lm/cpp/beam_search_causal_lm.cpp

```diff
@@ -0,0 +1,84 @@
+    tokenizer.infer();
+    return {tokenizer.get_tensor("input_ids"), tokenizer.get_tensor("attention_mask")};
+}
+
+std::string detokenize(ov::InferRequest& detokenizer, const std::vector<int64_t>& tokens) {
+    constexpr size_t BATCH_SIZE = 1;
+    ov::Tensor inp = detokenizer.get_input_tensor();
+    inp.set_shape({BATCH_SIZE, tokens.size()});
+    for (size_t idx = 0; idx < tokens.size(); ++idx) {
+        inp.data<int64_t>()[idx] = tokens.at(idx);
+    }
+    detokenizer.infer();
+    return openvino_extensions::unpack_strings(detokenizer.get_output_tensor()).front();
+}
+}
+
+int main(int argc, char* argv[]) try {
+    if (argc != 5) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0]
+            + " <openvino_model.xml> <tokenizer.xml> <detokenizer.xml> '<prompt>'");
+    }
+    ov::Core core;
+    core.add_extension(USER_OV_EXTENSIONS_PATH);  // USER_OV_EXTENSIONS_PATH is defined in root CMakeLists.txt
+    auto [input_ids, mask] = tokenize(core.compile_model(argv[2], "CPU").create_infer_request(), argv[4]);
+    ov::InferRequest detokenizer = core.compile_model(argv[3], "CPU").create_infer_request();
+    ov::InferRequest ireq = core.compile_model(argv[1], "CPU").create_infer_request();
+    ireq.set_tensor("input_ids", input_ids);
+    ireq.set_tensor("attention_mask", mask);
+    ov::Tensor position_ids = ireq.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    ireq.get_tensor("beam_idx").set_shape({1});
+    ireq.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+    Parameters parameters;
+    const int64_t* prompt_data = input_ids.data<const int64_t>();
+    parameters.prompt = std::vector<int64_t>{prompt_data, prompt_data + input_ids.get_size()};
+    GroupBeamSearcher group_beam_searcher{parameters};
+    std::vector<int64_t> next_tokens;
+    std::vector<int32_t> next_beams;
+    for (size_t length_count = 0; length_count < parameters.max_new_tokens; ++length_count) {
+        ireq.infer();
+        std::tie(next_tokens, next_beams) = group_beam_searcher.process(ireq.get_tensor("logits"));
+        if (next_tokens.empty()) {
+            break;
+        }
+        size_t batch_size = next_tokens.size();
+        ireq.set_tensor("input_ids", ov::Tensor{ov::element::i64, {batch_size, 1}, next_tokens.data()});
+        ov::Tensor attention_mask = ireq.get_tensor("attention_mask");
+        ov::Shape mask_shape{batch_size, attention_mask.get_shape().at(1) + 1};
+        attention_mask.set_shape(mask_shape);
+        std::fill_n(attention_mask.data<int64_t>(), shape_size(mask_shape), 1);
+        position_ids.set_shape({batch_size, 1});
+        std::fill_n(position_ids.data<int64_t>(), batch_size, mask_shape.at(1) - 1);
+        ireq.set_tensor("beam_idx", ov::Tensor{ov::element::i32, {batch_size}, next_beams.data()});
```

Comment:
Please provide description in comment what's happening here.

Reply:
Updated